### PR TITLE
Mat body force

### LINKED ISSUE #27353: add MatBodyForce
### Motivation

Body force that uses material property could be useful for some simulations.

### Design

Basically reproduce `MaskedBodyForce` in the `phase_field` modules. Do something simplier, but similar, for an AD version.

### Impact

More flexibility.

### PR DESCRIPTION
**Pull request by tophmatthews**
Closes #27353

### CODE REVIEW

**Comment by dschwen** [comment]
Is there anything that prevents the use of templating to combine the AD and non-AD versions?

**Comment by tophmatthews** [comment]
> Is there anything that prevents the use of templating to combine the AD and non-AD versions?

This was one of those times I thought it was so much easier to have separate objects due to `DerivativeMaterialInterface`.

**Comment by tophmatthews** [comment]
Also, should I deprecate `MaskedBodyForce`? I didn't `git mv` that class since `mask` didn't much make sense outside of that kernel, and it felt weird to make a new kernel with deprecations.

**Comment by dschwen** [comment]
> > Is there anything that prevents the use of templating to combine the AD and non-AD versions?
> 
> This was one of those times I thought it was so much easier to have separate objects due to `DerivativeMaterialInterface`.

Don't worry, I now just how to overengineer this!!

**Comment by dschwen** [comment]
> Also, should I deprecate `MaskedBodyForce`? I didn't `git mv` that class since `mask` didn't much make sense outside of that kernel, and it felt weird to make a new kernel with deprecations.

Well.. if you had moved it, it wouldn't be a "new" kernel, and we have a mechanism to rename parameters so that the old name is automatically deprecated. Either way, I don't want to make you jump through a bunch of hoops here. Let's just leave the PF kernel be for now.

**Comment by tophmatthews** [comment]
> Don't worry, I know just how to overengineer this!!

Cool! Happy to learn, got an example?

**Comment by tophmatthews** [comment]
I guess `test/src/MatKernel.C` should be deprecated too...

**Comment by dschwen** [comment]
I took the liberty to push a commit to refactor this a bit.

**Comment by tophmatthews** [comment]
whoa...badass.

**Comment by tophmatthews** [comment]
is `ADMatBodyForce.md` still needed?

File: framework/src/kernels/MatBodyForce.C [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  params.addRequiredParam<MaterialPropertyName>("material_property",
+                                                "Material property defining the property");
+  return params;
+}
+
+template <bool is_ad, class Parent>
+MatBodyForceTempl<is_ad, Parent>::MatBodyForceTempl(const InputParameters & parameters)
+  : Parent(parameters),
+    _property(this->template getGenericMaterialProperty<Real, is_ad>("material_property")),
+    _v_name(_var.name())
```
Comment (tophmatthews) [inline framework/src/kernels/MatBodyForce.C:31]: Should this be in the nonad constructor?

File: framework/include/kernels/MatBodyForce.h [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  static InputParameters validParams();
+
+  MatBodyForceTempl(const InputParameters & parameters);
+
+protected:
+  virtual GenericReal<is_ad> computeQpResidual() override;
+
+  const GenericMaterialProperty<Real, is_ad> & _property;
+
+  /// name of the nonlinear variable (needed to retrieve the derivative material properties)
```
Comment (tophmatthews) [inline framework/include/kernels/MatBodyForce.h:37]: Same here, should this be in the nonad constructor?

File: test/tests/kernels/body_force/ad_mat_forcing_function_test.i [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  []
+[]
+
+[Kernels]
+  [diff]
+    type = ADDiffusion
+    variable = u
+  []
+  [forcing]
+    type = ADMatBodyForce
```
Comment (tophmatthews) [inline test/tests/kernels/body_force/ad_mat_forcing_function_test.i:33]: Probably can delete this input file and just pass `ADParsedMaterial` and `ADMatBodyForce` via cliargs

File: framework/src/kernels/MatBodyForce.C (hunk 1 above)
Comment (dschwen) [reply framework/src/kernels/MatBodyForce.C]: No, it doesn't have to be

File: framework/include/kernels/MatBodyForce.h (hunk 2 above)
Comment (dschwen) [reply framework/include/kernels/MatBodyForce.h]: I'll push a fix

**Comment by dschwen** [comment]
> is `ADMatBodyForce.md` still needed?

I can never remember. Probably not.

**Review by dschwen — APPROVED** [review]
This generally looks fine to me. Since I pushed a commit to refactor this, we should get another independent review.

File: framework/src/kernels/MatBodyForce.C [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "MatBodyForce.h"
+#include "Function.h"
```
Comment (dschwen) [inline framework/src/kernels/MatBodyForce.C:11]: Doesn't seem to be used (and neither is the forward declaration in the header)

Comment (tophmatthews) [reply framework/src/kernels/MatBodyForce.C:11]: Yeah, copy paste error

**Comment by tophmatthews** [comment]
Had to add back the full ad test input file since I couldn't suppress `coupled_variables`. Guess I could have supplied that via cliargs for non-ad test, but then you get input files that don't work by themselves.

File: framework/src/kernels/MatBodyForce.C [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+MatBodyForce::computeQpJacobian()
+{
+  return _dpropertydv[_qp] * BodyForce::computeQpResidual() * _phi[_j][_qp];
+}
+
+Real
+MatBodyForce::computeQpOffDiagJacobian(unsigned int jvar)
+{
+  const unsigned int cvar = mapJvarToCvar(jvar);
+  return (*_dpropertydarg[cvar])[_qp] * BodyForce::computeQpResidual() * _phi[_j][_qp];
```
Comment (tophmatthews) [inline framework/src/kernels/MatBodyForce.C:66]: Isn't there supposed to be an if statement check here if jvar is the one actually provided in the material property @dschwen ?

**Review by loganharbour — APPROVED** [review]

File: framework/src/kernels/MatBodyForce.C (hunk 5 above)
Comment (loganharbour) [reply framework/src/kernels/MatBodyForce.C:66]: We should only be calling this for the entries that have coupling

Comment (tophmatthews) [reply framework/src/kernels/MatBodyForce.C:66]: Yeah, I found other examples that look exactly like this, I think I was wrong in my assumption.

**Pull request merged**